CMake: use build type as a profile selection

0xc0170 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Using `CMAKE_BUILD_TYPE` as a profile selector. We provide 3 build types (Mbed OS profiles): Debug, Release and Develop. 

If this is accepted the user should configure CMake with `cmake .. -G Ninja -D CMAKE_BUILD_TYPE=${PROFILE}` instead of `cmake .. -G Ninja -D MBED_PROFILE=${PROFILE}`.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [x] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

0xc0170 committed:
CMake: using CMAKE_BUILD_TYPE for choosing a profile

Select a profile using `-DCMAKE_BUILD_TYPE=<profile>`.
We support 3 profiles: Debug, Develop and Release.

hugueskamba commented on cmake/profiles/debug.cmake:
```diff
@@ -1,6 +1,11 @@
 # Copyright (c) 2020 ARM Limited. All rights reserved.
 # SPDX-License-Identifier: Apache-2.0
 
+set(CMAKE_BUILD_TYPE "Release"   
```
Did you mean `"Debug"`?

0xc0170 replied:
Bad rebase, fixing

hugueskamba commented on cmake/profile.cmake:
```diff
@@ -1,4 +1,19 @@
+else()
+  set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Build Type of the project.")
+  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "${MBED_BUILD_TYPES}")
+  if(NOT CMAKE_BUILD_TYPE IN_LIST MBED_BUILD_TYPES)
```
Is the check case sensitive?

0xc0170 replied:
I'll lower the types and what comes from the tools

0xc0170 replied:
fixed (case insensitive in checking now)

hugueskamba commented on cmake/profile.cmake:
```diff
@@ -1,4 +1,19 @@
+if(multi_config)
+  set(CMAKE_CONFIGURATION_TYPES "${MBED_BUILD_TYPES}" CACHE STRING "list of supported configuration types" FORCE)
+else()
+  set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Build Type of the project.")
```
```suggestion
  set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Build type of the project.")
```

0xc0170 replied:
fixed

hugueskamba commented on cmake/profile.cmake:
```diff
@@ -1,4 +1,19 @@
+set(MBED_BUILD_TYPES Debug Release Develop)
+get_property(multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+if(multi_config)
+  set(CMAKE_CONFIGURATION_TYPES "${MBED_BUILD_TYPES}" CACHE STRING "list of supported configuration types" FORCE)
```
```suggestion
  set(CMAKE_CONFIGURATION_TYPES "${MBED_BUILD_TYPES}" CACHE STRING "List of supported build types" FORCE)
```
For consistency.

0xc0170 replied:
fixed

hugueskamba commented on cmake/profile.cmake:
```diff
@@ -1,4 +1,22 @@
+if(multi_config)
+  set(CMAKE_CONFIGURATION_TYPES "${MBED_BUILD_TYPES}" CACHE STRING "List of supported build types" FORCE)
+else()
+  set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Build type of the project.")
```
Does this line mean that the default build type is `Release`? If so, should it not be `Develop` as with Mbed CLI?

0xc0170 replied:
> Set still needs a fix, will update

hold on review until updated,

0xc0170 committed:
CMake build type: set default Develop profile

Adding also comments for multiconfig/property set.

0xc0170 committed:
CMake profile: fix alignment

hugueskamba commented on cmake/profile.cmake:
```diff
@@ -1,4 +1,30 @@
 # SPDX-License-Identifier: Apache-2.0
 
-include(${MBED_ROOT}/cmake/profiles/${MBED_PROFILE}.cmake)
+# Set Develop as our default type if no other selected 
```
```suggestion
# Set the default build type if none is specified
```
I would not make the comment dependent on the select default (in case it changes).

hugueskamba commented on cmake/profile.cmake:
```diff
@@ -1,4 +1,30 @@
+else()
+    # Set the possible values of build type for cmake-gui
+    set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "${MBED_BUILD_TYPES}")
+    if(NOT LOWERCASE_CMAKE_BUILD_TYPE IN_LIST LOWERCASE_MBED_BUILD_TYPES)
```
Maybe a empty line between the code at line 24 and 25 so it is clear that the comment only applies to line 24.

hugueskamba commented on cmake/profile.cmake:
```diff
@@ -1,4 +1,30 @@
+
+get_property(multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+if(multi_config)
+    # multi configuration generators do not support build type, providing a list only
```
```suggestion
    # Provide only a list as multi configuration generators do not support build type
```

0xc0170 committed:
Travis: use build type for cmake

0xc0170 committed:
CMake profile: fix comments

0xc0170 committed:
CMake toolchain: clear flags to be able to overwrite them

As Mbed uses different optimizations levels, we should be able to overwrite this via toolchain file.

0xc0170 committed:
CMake GCC ARM: use nosys specs for linker

Use retarged version (nosys.specs) for Mbed OS. To resolve an issue with debug/develop profiles.
Otherwise we get kill or getpid errors from linker, as we use nosys lib but no specs defined.

0xc0170 commented on cmake/toolchain.cmake:
```diff
@@ -15,3 +15,17 @@ set(CMAKE_CXX_COMPILER_WORKS TRUE)
+
+# Clear FLAGS for all languages
+# Mbed OS uses different init flags (optimization, debug symbols) 
+set(CMAKE_C_FLAGS_DEBUG "" CACHE STRING "" FORCE)
```
For reviewers, as our profile flags differ (android for instance does the same in their toolchain file) we clear the the flags after init (happens on line 17th in enable_language). Just making sure we have flags we set.

hugueskamba commented on cmake/profile.cmake:
```diff
@@ -1,4 +1,31 @@
+# Set the default build type if none is specified
+if(NOT CMAKE_BUILD_TYPE)
+    set(CMAKE_BUILD_TYPE "Develop" CACHE
+        STRING "The type of build" FORCE)
```
```suggestion
        STRING "The build type" FORCE
   )
```

hugueskamba commented on cmake/profile.cmake:
```diff
@@ -1,4 +1,31 @@
+        STRING "The type of build" FORCE)
+endif()
+
+# Mapping CMAKE_BUILD_TYPE into MBED_BUILD_TYPES, as we understand only 3 profiles
```
I do not think this comment belongs on this line.
It should probably go after the empty line on line 17 (right before `get_property(multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)`).

All you are doing there is setting a variable.

0xc0170 replied:
reviewing now and rebasing to add your 2 suggestion

hugueskamba commented on cmake/toolchain.cmake:
```diff
@@ -15,3 +15,17 @@ set(CMAKE_CXX_COMPILER_WORKS TRUE)
 # Project setup
 enable_language(C CXX ASM)
+
+# Clear FLAGS for all languages
```
```suggestion
# Clear toolchains options for all languages as Mbed OS uses 
# different initialisation options (such as for optimization and debug symbols)
```

hugueskamba commented on cmake/toolchains/GCC_ARM.cmake:
```diff
@@ -15,6 +15,7 @@ list(APPEND link_options
     "-lgcc"
     "-lnosys"
     "-Wl,--end-group"
+    "-specs=nosys.specs"
```
Why was this added?

0xc0170 replied:
It should be described in the commit message. lib nosys but that is not sufficient as there are specs as well and did not reflect the library selection. There are librdimon.a or libnosys.a. as we use already `-lnosys`, we should as well use specs.

If we remove this, we get linker complaining about not finding _kill and other functions (2 out of 3 profiles were failing).

There are 3 specs available rdimon, nosys or nano by GCC ARM.

hugueskamba replied:
Sorry I did not look at individual commits.

Was the same option added by Mbed CLI? If not, are we masking something else by including this flag?

0xc0170 replied:
Some exporters had it but I could not find it in tools. Please fetch this PR and compare flags between this and master, to be certain

hugueskamba replied:
I investigated the cause of the failure because all three profiles were working fine prior to your changes.

If the `RelWithDebInfo` profile is used alongside Mbed OS `Develop` profile the linker completes successfully as long as `set(CMAKE_C_FLAGS_RELWITHDEBINFO "" CACHE STRING "" FORCE)` is commented out. It fails exactly as you described when that line is re-enabled.

However, using Mbed OS `Develop` profile on its own with that `set(CMAKE_C_FLAGS_RELWITHDEBINFO "" CACHE STRING "" FORCE)` commented out still fails.

The `RelWithDebInfo` must add something in the C flags to ensure we do not get that linker failure. However, inspecting the system information (with `cmake --system-information information.txt`) does not show `-specs=nosys.specs`.

My main concern was that somehow printing on the console gets broken with this link option. However, I tested that it is still possible to print on the console.

I do not have an objection to add the new option.

0xc0170 replied:
Thanks !

0xc0170 committed:
CMake toolchain: fix options comment

Co-authored-by: Hugues Kamba <41612201+hugueskamba@users.noreply.github.com>

0xc0170 committed:
CMake: Profile comments update

Co-authored-by: Hugues Kamba <41612201+hugueskamba@users.noreply.github.com>

hugueskamba commented on cmake/toolchain.cmake:
```diff
@@ -15,3 +15,18 @@ set(CMAKE_CXX_COMPILER_WORKS TRUE)
+
+# Clear toolchains options for all languages as Mbed OS uses 
+# different initialisation options (such as for optimization and debug symbols)
+# Mbed OS uses different init flags (optimization, debug symbols) 
```
Line 21 now needs to be removed.
```suggestion

```

0xc0170 committed:
Update cmake/toolchain.cmake

Co-authored-by: Hugues Kamba <41612201+hugueskamba@users.noreply.github.com>
